comit/src/hbit.rs

```diff
@@ -124,23 +133,32 @@ async fn watch_ledger(
     connector: Arc<Cache<BitcoindConnector>>,
     params: Params,
     co: &Co<anyhow::Result<Event>>,
+    start_of_swap: NaiveDateTime,
 ) -> anyhow::Result<()> {
-    let funded = connector.htlc_funded(&params, start_of_swap).await?;
-    co.yield_(Event::Funded(funded)).await;
+    let funded = connector.wait_for_funded(&params, start_of_swap).await?;
+    co.yield_(Ok(Event::Funded(funded.clone()))).await;
 
-    let redeemed = connector.htlc_redeemed(&params, &funded, start_of_swap);
-    let refunded = connector.htlc_refunded(&params, &funded, start_of_swap);
+    match funded {
+        Funded::Correctly { location, .. } => {
+            let redeemed = connector.wait_for_redeemed(&params, location, start_of_swap);
+            let refunded = connector.wait_for_refunded(&params, location, start_of_swap);
 
-    match future::try_select(redeemed, refunded).await {
-        Ok(Either::Left((redeemed, _))) => {
-            co.yield_(Event::Redeemed(redeemed.clone())).await;
-        }
-        Ok(Either::Right((refunded, _))) => {
-            co.yield_(Event::Refunded(refunded.clone())).await;
+            match future::try_select(redeemed, refunded).await {
+                Ok(Either::Left((redeemed, _))) => {
+                    co.yield_(Ok(Event::Redeemed(redeemed.clone()))).await;
+                }
+                Ok(Either::Right((refunded, _))) => {
+                    co.yield_(Ok(Event::Refunded(refunded.clone()))).await;
+                }
+                Err(either) => {
+                    let (error, _other_future) = either.factor_first();
+                    return Err(error);
+                }
+            }
         }
-        Err(either) => {
-            let (error, _other_future) = either.factor_first();
-            return Err(error);
+        Funded::Incorrectly { .. } => {
+            // FIXME: Is this what we want?
+            tracing::info!("Htlc was incorrectly funded, aborting watcher");
```

Comment:
I would say no. This part of the code is not about making decisions, it is about observing.

Redeem and refund can still happen, even if incorrectly funded :)

Reply:
Cool, thanks. Will fix.